### PR TITLE
Add functionality for domain groups

### PR DESCRIPTION
**Pull request by brennenawana**
We want to be able authorize a user based on their email domain.
This functionality follows the same a pattern as the way `groups` are handled. _(see examples/underpants.groups.json)_

Related: https://github.com/playdots/underpants-eb/pull/13

### CODE REVIEW

File: config/context.go [hunk 1]
```diff
@@ -74,3 +85,23 @@ func (c *Context) UserMemberOfAny(email string, groups []string) bool {
 
 	return false
 }
+
+// DomainMemberOfAny determines if an email domain belongs to any domain groups.
+func (c *Context) DomainMemberOfAny(domain string, domainGroups []string) bool {
+	if !c.HasDomainGroups() {
```
Comment (alexhornbake) [inline config/context.go]: so if I forget to add the domain group to the config I believe this defaults to allowing any domain?

File: proxy/backend.go [hunk 2]
```diff
@@ -78,6 +78,19 @@ func (b *Backend) serveHTTPProxy(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	email := strings.Split(u.Email, "@")
```
Comment (alexhornbake) [inline proxy/backend.go]: I think we should be more defensive here, and assert our expectation that an email will only ever contain one "@"

File: config/context_test.go [hunk 3]
```diff
@@ -42,3 +42,41 @@ func TestUserMemberOfAny(t *testing.T) {
...
+	}
+
+	for _, test := range tests {
+		if ctx.DomainMemberOfAny(test.Domain, test.DomainGroups) != test.Expected {
+			t.Fatalf("%s member of any of %s should have been %t",
+				test.Domain,
+				strings.Join(test.DomainGroups, ","),
+				test.Expected)
+		}
+	}
```
Comment (alexhornbake) [inline config/context_test.go]: would be good to have a test for when DomainGroups are not supplied, and that no domains are allowed.

File: config/context.go (hunk 1 above)
Comment (brennenawana) [reply config/context.go]: I went through 4 likely scenarios, this is how they stack up:

#### Scenario 1: Config is set up with`domain-groups` && route has `allowed-domain-groups` set.
- Behaves as expected, only authorizes allowed domains for that route

#### Scenario 2: Config is set up with`domain-groups` but route does not have `allowed-domain-groups` set.
- Authorization for that route fails for all domains

#### Scenario 3: Config is not set up with`domain-groups` but is configured with `oauth: domain`.
- Authorization for that route is determined by `oauth:domain`, which in our case is set by ENV['OAUTH_DOMAIN'].

#### Scenario 4: Config is not set up with `domain-groups` or `oauth:domain`.
- All routes fail Authorization for all domains

Comment (alexhornbake) [reply config/context.go]: Sounds good. :+1: Lets add test coverage to verify.

**Pull request merged**